### DISCUSSION TITLE
iPXE AMD SEV (Confidential Computing) support?

### DISCUSSION THREAD

**Discussion by ajyoung-oracle** [ipxe/ipxe#487] [Q&A]
We have been trying to use iPXE with AMD SEV (Confidential Computing) feature enabled in a VM and are getting crashes (from OVMF/iPXE).

I have seen it mentioned in a few places around the net that "iPXE does not yet support AMD SEV".

Is this true? If so, what is not supported exactly? And, what, if any, plans are there to support AMD SEV?

Thanks in advance for any help/info/comments.

**Comment by NiKiZe** [comment]
Do you have any info on the crashes?

If this is something that you need specific support for you should reach out to get commercial support and get the feature supported.

**Comment by mcb30** [comment]
> We have been trying to use iPXE with AMD SEV (Confidential Computing) feature enabled in a VM and are getting crashes (from OVMF/iPXE).
> 
> I have seen it mentioned in a few places around the net that "iPXE does not yet support AMD SEV".
> 
> Is this true? If so, what is not supported exactly? And, what, if any, plans are there to support AMD SEV?

The entire purpose of SEV, as I understand it, is to allow basically unmodified binaries to run inside the "secure" virtual machine.  It sounds as though there is therefore a bug in SEV if it fails to support iPXE.  :slightly_smiling_face: 

There are no plans at present to do any specific work to fix any interoperability issues between SEV and iPXE.  Having seen the quagmire of Confidential Computing patches for OVMF currently going around the EDK2 mailing list: please be aware that external patches relating to SEV are very unlikely to be merged into iPXE unless they are very small (less than 100 lines), obviously correct, and easily tested.  In all honesty, it is unlikely that any work on this feature will happen unless there is commercial funding to do so.

**Reply by ajyoung-oracle** [reply]
Thanks for the replies. 

As requested, below is some debug info on the crashes. 

Appears to be crashing in iPXE as soon as iPXE efi_snp (SNPDEV) code starts up running.
Per the ovmf debug log, 1af41000.efidrv was the last driver loaded before crash (and you can see 1af41000.efidrv flagged in the OVMF exception data).
Also, if we set romfile="" (to avoid loading iPXE OPROM), it boots up (from the PV disk).

btw: It seems to be somewhat well-known in libvirt docs that iPXE "doesn't work with SEV" from these sites on the net:
https://libvirt.org/kbase/launch_security_sev.html (section on virtio-net)
https://github.com/libvirt/libvirt/blob/master/docs/kbase/launch_security_sev.rst

If anyone has any suggestions/recommendations on how to fix this, please let me know. Thanks in advance!

1af41000.efidrv iPXE target built from top commit: 2265a65191d76ce367913a61c97752ab88ab1a59
DEBUG ARGS = "iscsi:3,scsi:3,tls:3,sanboot:3,x509:3,certstore:3,privkey:3,efi_acpi:3,efi_autoboot:3
,efi_autoexec:3,efi_block:3,efi_bofm:3,efi_cachedhcp:3,efi_console:3,efi_debug:3,efi_download:3,efi_driv
er:3,efi_entropy:3,efi_fbcon:3,efi_fdt:3,efi_file:3,efi_guid:3,efi_hii:3,efi_init:3,efi_local:3,efi_null
:3,efi_path:3,efi_pci:3,efi_pxe:3,efi_reboot:3,efi_smbios:3,efi_snp:3,efi_snp_hii:3,efi_strings:3,efi_ti
me:3,efi_timer:3,efi_uaccess:3,efi_umalloc:3,efi_usb:3,efi_utils:3,efi_veto:3,efi_watchdog:3,efi_wrap:3,
efidrvprefix:3,efiprefix:3"

OVMF built from top commit: b8649cf2a3e673a4a8cb6c255e394b354b771550

QEMU args:
/usr/bin/qemu-system-x86_64 --no-user-config \
-machine type=q35,accel=kvm,usb=on,dump-guest-core=off -rtc base=utc \
-cpu host,+host-phys-bits -smp cpus=128,cores=64,threads=2,sockets=1 -m 256G
\
-drive
if=pflash,format=raw,unit=0,file=/usr/share/OVMF/OVMF_CODE.pure-efi.fd,readonl
y \
-drive if=pflash,format=raw,unit=1,file=/root/kvm/OVMF_VARS.pure-efi.fd \
-drive file=/root/kvm/mysql7.qcow2,if=none,id=disk0,format=qcow2 \
-device virtio-scsi-pci,id=scsi,disable-legacy=on,iommu_platform=true \
-device scsi-hd,drive=disk0 \
-device virtio-rng-pci,disable-legacy=on,iommu_platform=true \
-object sev-guest,id=sev0,cbitpos=51,reduced-phys-bits=1 \
-machine memory-encryption=sev0 \
-nodefaults \
-global ICH9-LPC.disable_s3=1 \
-device virtio-net-pci,mac=$(cat
/sys/class/net/macvtap1/address),netdev=hostnet0,disable-legacy=on,iommu_platf
orm=on,romfile=/usr/share/ipxe/1af41000.efirom \
-netdev tap,fd=3,id=hostnet0 3<>/dev/tap$(cat
/sys/class/net/macvtap1/ifindex) \
-nographic -serial telnet:127.0.0.1:4556,server 

Serial console log:

EFIPCI 0000:00:1d.0 (8086:2934 class 0c0300) has no driver
EFIPCI 0000:00:1d.0 (8086:2934 class 0c0300) has no driver
EFIPCI 0000:00:1d.0 (8086:2934 class 0c0300) has no driver
EFIPCI 0000:00:1d.0 (8086:2934 class 0c0300) has no driver
EFIPCI 0000:00:1d.1 (8086:2935 class 0c0300) has no driver
EFIPCI 0000:00:1d.1 (8086:2935 class 0c0300) has no driver
EFIPCI 0000:00:1d.1 (8086:2935 class 0c0300) has no driver
EFIPCI 0000:00:1d.1 (8086:2935 class 0c0300) has no driver
EFIPCI 0000:00:1d.2 (8086:2936 class 0c0300) has no driver
EFIPCI 0000:00:1d.2 (8086:2936 class 0c0300) has no driver
EFIPCI 0000:00:1d.2 (8086:2936 class 0c0300) has no driver
EFIPCI 0000:00:1d.2 (8086:2936 class 0c0300) has no driver
EFIPCI 0000:00:1d.7 (8086:293a class 0c0320) has no driver
EFIPCI 0000:00:1d.7 (8086:293a class 0c0320) has no driver
EFIPCI 0000:00:1f.2 (8086:2922 class 010601) has no driver
EFIPCI 0000:00:1f.2 (8086:2922 class 010601) has no driver
EFIPCI 0000:00:1f.2 (8086:2922 class 010601) has no driver
EFIPCI 0000:00:1f.2 (8086:2922 class 010601) has no driver
EFIPCI 0000:00:1f.2 (8086:2922 class 010601) has no driver
EFIPCI 0000:00:01.0 (1af4:1048 class 010000) has no driver
EFIPCI 0000:00:01.0 (1af4:1048 class 010000) has no driver
EFIPCI 0000:00:01.0 (1af4:1048 class 010000) has no driver
EFIPCI 0000:00:01.0 (1af4:1048 class 010000) has no driver
EFIPCI 0000:00:1d.0 (8086:2934 class 0c0300) has no driver
EFIPCI 0000:00:1d.0 (8086:2934 class 0c0300) has no driver
EFIPCI 0000:00:1d.1 (8086:2935 class 0c0300) has no driver
EFIPCI 0000:00:1d.1 (8086:2935 class 0c0300) has no driver
EFIPCI 0000:00:1d.2 (8086:2936 class 0c0300) has no driver
EFIPCI 0000:00:1d.2 (8086:2936 class 0c0300) has no driver
EFIPCI 0000:00:1d.7 (8086:293a class 0c0320) has no driver
EFIPCI 0000:00:1d.7 (8086:293a class 0c0320) has no driver
EFIPCI 0000:00:00.0 (8086:29c0 class 060000) has no driver
EFIPCI 0000:00:01.0 (1af4:1048 class 010000) has no driver
EFIPCI 0000:00:01.0 (1af4:1048 class 010000) has no driver
EFIPCI 0000:00:02.0 (1af4:1044 class 00ff00) has no driver
EFIPCI 0000:00:03.0 (1af4:1041 class 020000) has driver "virtio-net"
<0-Y>EFIDRV PciRoot(0x0)/Pci(0x3,0x0) has driver "PCI"
<0-Y>EFIDRV PciRoot(0x0)/Pci(0x3,0x0) DRIVER_START
EFIPCI 0000:00:03.0 (1af4:1041 class 020000) has driver "virtio-net"
EFIPCI 0000:00:03.0 found range [c0000000,c0100000) -> [c0000000,c0100000)
EFIPCI 0000:00:03.0 found range [4800000000,4800100000) -> [4800000000,4800100000)
EFIPCI 0000:00:03.0 4800008000+38 -> 4800008000
EFIPCI 0000:00:03.0 found range [c0000000,c0100000) -> [c0000000,c0100000)
EFIPCI 0000:00:03.0 found range [4800000000,4800100000) -> [4800000000,4800100000)
EFIPCI 0000:00:03.0 4800009000+1 -> 4800009000
EFIPCI 0000:00:03.0 found range [c0000000,c0100000) -> [c0000000,c0100000)
EFIPCI 0000:00:03.0 found range [4800000000,4800100000) -> [4800000000,4800100000)
EFIPCI 0000:00:03.0 480000a000+c -> 480000a000
PRIVKEY has no private key
<0-Y>EFIDEV PciRoot(0x0)/Pci(0x3,0x0) added child<0-Y> PciRoot(0x0)/Pci(0x3,0x0)/MAC(5254FFFFEE1B,0x1)
IFR 0x7faf0710 string 0x0001 is "iPXE (52:54:ff:ff:ee:1b)"
IFR 0x7faf0710 string 0x0002 is "Configure iPXE"
IFR 0x7faf0710 formset title 0x0001 help 0x0002
00000000 : 0e b7 31 0a 60 1a 43 51-d2 da a0 d6 ef 30 00 b9 : ..1.`.CQ.....0..
00000010 : c8 8b 01 00 02 00 02 71-99 03 93 45 85 04 4b b4 : .......q...E..K.
00000020 : 5e 32 eb 83 26 04 0e 46-97 8e 5c f7 a5 93 45 af : ^2..&..F..\...E.
00000030 : 1f 66 a8 2a a1 9c b1                            : .f.*...         
IFR 0x7faf0710 GUID class 0x04
00000037 : 5f 15 35 17 0b 0f a0 87-93 41 b2 66 53 8c 38 af : _.5......A.fS.8.
00000047 : 48 ce 03 04 00                                  : H....           
IFR 0x7faf0710 GUID subclass 0x03
0000004c : 5f 15 35 17 0b 0f a0 87-93 41 b2 66 53 8c 38 af : _.5......A.fS.8.
0000005c : 48 ce 04 03 00                                  : H....           
IFR 0x7faf0710 name/value store 0x0001
00000061 : 25 14 01 00 c9 5f 12 f7-98 ea 06 3c 50 15 73 e5 : %...._.....<P.s.
00000071 : e0 ed 25 19                                     : ..%.            
IFR 0x7faf0710 name/value store 0x0001 title 0x0001
00000075 : 01 86 01 00 01 00                               : ......          
IFR 0x7faf0710 string 0x0003 is "iPXE"
IFR 0x7faf0710 string 0x0004 is "Firmware product name"
IFR 0x7faf0710 string 0x0005 is "Name"
IFR 0x7faf0710 text prompt 0x0005 help 0x0004 text 0x0003
0000007b : 03 08 05 00 04 00 03 00                         : ........        
IFR 0x7faf0710 string 0x0006 is "1.0.0+"
IFR 0x7faf0710 string 0x0007 is "Firmware version"
IFR 0x7faf0710 string 0x0008 is "Version"
IFR 0x7faf0710 text prompt 0x0008 help 0x0007 text 0x0006
00000083 : 03 08 08 00 07 00 06 00                         : ........        
IFR 0x7faf0710 string 0x0009 is "virtio-net"
IFR 0x7faf0710 string 0x000a is "Firmware driver"
IFR 0x7faf0710 string 0x000b is "Driver"
IFR 0x7faf0710 text prompt 0x000b help 0x000a text 0x0009
0000008b : 03 08 0b 00 0a 00 09 00                         : ........        
IFR 0x7faf0710 string 0x000c is "0000:00:03.0"
IFR 0x7faf0710 string 0x000d is "Hardware device"
IFR 0x7faf0710 string 0x000e is "Device"
IFR 0x7faf0710 text prompt 0x000e help 0x000d text 0x000c
00000093 : 03 08 0e 00 0d 00 0c 00                         : ........        
IFR 0x7faf0710 string 0x000f is "mtu"
IFR 0x7faf0710 string 0x0010 is "MTU"
IFR 0x7faf0710 string 0x0011 is "http://ipxe.org/cfg/mtu"
IFR 0x7faf0710 string prompt 0x0010 help 0x0011 question 0x001a varstore 0x0001:0x000f
0000009b : 1c 10 10 00 11 00 1a 00-01 00 0f 00 00 00 ff 00 : ................
IFR 0x7faf0710 string 0x0012 is "gateway"
IFR 0x7faf0710 string 0x0013 is "Default gateway"
IFR 0x7faf0710 string 0x0014 is "http://ipxe.org/cfg/gateway"
IFR 0x7faf0710 string prompt 0x0013 help 0x0014 question 0x0003 varstore 0x0001:0x0012
000000ab : 1c 10 13 00 14 00 03 00-01 00 12 00 00 00 ff 00 : ................
IFR 0x7faf0710 string 0x0015 is "ip"
IFR 0x7faf0710 string 0x0016 is "IP address"
IFR 0x7faf0710 string 0x0017 is "http://ipxe.org/cfg/ip"
IFR 0x7faf0710 string prompt 0x0016 help 0x0017 question 0xaf02 varstore 0x0001:0x0015
000000bb : 1c 10 16 00 17 00 02 af-01 00 15 00 00 00 ff 00 : ................
IFR 0x7faf0710 string 0x0018 is "netmask"
IFR 0x7faf0710 string 0x0019 is "Subnet mask"
IFR 0x7faf0710 string 0x001a is "http://ipxe.org/cfg/netmask"
IFR 0x7faf0710 string prompt 0x0019 help 0x001a question 0x0001 varstore 0x0001:0x0018
000000cb : 1c 10 19 00 1a 00 01 00-01 00 18 00 00 00 ff 00 : ................
IFR 0x7faf0710 string 0x001b is "dns"
IFR 0x7faf0710 string 0x001c is "DNS server"
IFR 0x7faf0710 string 0x001d is "http://ipxe.org/cfg/dns"
IFR 0x7faf0710 string prompt 0x001c help 0x001d question 0x0006 varstore 0x0001:0x001b
000000db : 1c 10 1c 00 1d 00 06 00-01 00 1b 00 00 00 ff 00 : ................
IFR 0x7faf0710 string 0x001e is "dnssl"
IFR 0x7faf0710 string 0x001f is "DNS search list"
IFR 0x7faf0710 string 0x0020 is "http://ipxe.org/cfg/dnssl"
IFR 0x7faf0710 string prompt 0x001f help 0x0020 question 0x0077 varstore 0x0001:0x001e
000000eb : 1c 10 1f 00 20 00 77 00-01 00 1e 00 00 00 ff 00 : .... .w.........
IFR 0x7faf0710 string 0x0021 is "domain"
IFR 0x7faf0710 string 0x0022 is "DNS domain"
IFR 0x7faf0710 string 0x0023 is "http://ipxe.org/cfg/domain"
IFR 0x7faf0710 string prompt 0x0022 help 0x0023 question 0x000f varstore 0x0001:0x0021
000000fb : 1c 10 22 00 23 00 0f 00-01 00 21 00 00 00 ff 00 : ..".#.....!.....
IFR 0x7faf0710 string 0x0024 is "filename"
IFR 0x7faf0710 string 0x0025 is "Boot filename"
IFR 0x7faf0710 string 0x0026 is "http://ipxe.org/cfg/filename"
IFR 0x7faf0710 string prompt 0x0025 help 0x0026 question 0x0043 varstore 0x0001:0x0024
0000010b : 1c 10 25 00 26 00 43 00-01 00 24 00 00 00 ff 00 : ..%.&.C...$.....
IFR 0x7faf0710 string 0x0027 is "next-server"
IFR 0x7faf0710 string 0x0028 is "TFTP server"
IFR 0x7faf0710 string 0x0029 is "http://ipxe.org/cfg/next-server"
IFR 0x7faf0710 string prompt 0x0028 help 0x0029 question 0xaf03 varstore 0x0001:0x0027
0000011b : 1c 10 28 00 29 00 03 af-01 00 27 00 00 00 ff 00 : ..(.).....'.....
IFR 0x7faf0710 string 0x002a is "root-path"
IFR 0x7faf0710 string 0x002b is "SAN root path"
IFR 0x7faf0710 string 0x002c is "http://ipxe.org/cfg/root-path"
IFR 0x7faf0710 string prompt 0x002b help 0x002c question 0x0011 varstore 0x0001:0x002a
0000012b : 1c 10 2b 00 2c 00 11 00-01 00 2a 00 00 00 ff 00 : ..+.,.....*.....
IFR 0x7faf0710 string 0x002d is "san-filename"
IFR 0x7faf0710 string 0x002e is "SAN filename"
IFR 0x7faf0710 string 0x002f is "http://ipxe.org/cfg/san-filename"
IFR 0x7faf0710 string prompt 0x002e help 0x002f question 0xafbc varstore 0x0001:0x002d
0000013b : 1c 10 2e 00 2f 00 bc af-01 00 2d 00 00 00 ff 00 : ..../.....-.....
IFR 0x7faf0710 string 0x0030 is "initiator-iqn"
IFR 0x7faf0710 string 0x0031 is "iSCSI initiator name"
IFR 0x7faf0710 string 0x0032 is "http://ipxe.org/cfg/initiator-iqn"
IFR 0x7faf0710 string prompt 0x0031 help 0x0032 question 0x00cb varstore 0x0001:0x0030
0000014b : 1c 10 31 00 32 00 cb 00-01 00 30 00 00 00 ff 00 : ..1.2.....0.....
IFR 0x7faf0710 string 0x0033 is "keep-san"
IFR 0x7faf0710 string 0x0034 is "Preserve SAN connection"
IFR 0x7faf0710 string 0x0035 is "http://ipxe.org/cfg/keep-san"
IFR 0x7faf0710 string prompt 0x0034 help 0x0035 question 0xaf08 varstore 0x0001:0x0033
0000015b : 1c 10 34 00 35 00 08 af-01 00 33 00 00 00 ff 00 : ..4.5.....3.....
IFR 0x7faf0710 string 0x0036 is "san-drive"
IFR 0x7faf0710 string 0x0037 is "SAN drive number"
IFR 0x7faf0710 string 0x0038 is "http://ipxe.org/cfg/san-drive"
IFR 0x7faf0710 string prompt 0x0037 help 0x0038 question 0xafbd varstore 0x0001:0x0036
0000016b : 1c 10 37 00 38 00 bd af-01 00 36 00 00 00 ff 00 : ..7.8.....6.....
IFR 0x7faf0710 string 0x0039 is "san-retries"
IFR 0x7faf0710 string 0x003a is "SAN retry count"
IFR 0x7faf0710 string 0x003b is "http://ipxe.org/cfg/san-retries"
IFR 0x7faf0710 string prompt 0x003a help 0x003b question 0xafbb varstore 0x0001:0x0039
0000017b : 1c 10 3a 00 3b 00 bb af-01 00 39 00 00 00 ff 00 : ..:.;.....9.....
IFR 0x7faf0710 string 0x003c is "skip-san-boot"
IFR 0x7faf0710 string 0x003d is "Do not boot from SAN device"
IFR 0x7faf0710 string 0x003e is "http://ipxe.org/cfg/skip-san-boot"
IFR 0x7faf0710 string prompt 0x003d help 0x003e question 0xaf09 varstore 0x0001:0x003c
0000018b : 1c 10 3d 00 3e 00 09 af-01 00 3c 00 00 00 ff 00 : ..=.>.....<.....
IFR 0x7faf0710 string 0x003f is "hostname"
IFR 0x7faf0710 string 0x0040 is "Host name"
IFR 0x7faf0710 string 0x0041 is "http://ipxe.org/cfg/hostname"
IFR 0x7faf0710 string prompt 0x0040 help 0x0041 question 0x000c varstore 0x0001:0x003f
0000019b : 1c 10 40 00 41 00 0c 00-01 00 3f 00 00 00 ff 00 : ..@.A.....?.....
IFR 0x7faf0710 string 0x0042 is "user-class"
IFR 0x7faf0710 string 0x0043 is "DHCP user class"
IFR 0x7faf0710 string 0x0044 is "http://ipxe.org/cfg/user-class"
IFR 0x7faf0710 string prompt 0x0043 help 0x0044 question 0x004d varstore 0x0001:0x0042
000001ab : 1c 10 43 00 44 00 4d 00-01 00 42 00 00 00 ff 00 : ..C.D.M...B.....
IFR 0x7faf0710 string 0x0045 is "vendor-class"
IFR 0x7faf0710 string 0x0046 is "DHCP vendor class"
IFR 0x7faf0710 string 0x0047 is "http://ipxe.org/cfg/vendor-class"
IFR 0x7faf0710 string prompt 0x0046 help 0x0047 question 0x003c varstore 0x0001:0x0045
000001bb : 1c 10 46 00 47 00 3c 00-01 00 45 00 00 00 ff 00 : ..F.G.<...E.....
IFR 0x7faf0710 string 0x0048 is "password"
IFR 0x7faf0710 string 0x0049 is "Password"
IFR 0x7faf0710 string 0x004a is "http://ipxe.org/cfg/password"
IFR 0x7faf0710 string prompt 0x0049 help 0x004a question 0xafbf varstore 0x0001:0x0048
000001cb : 1c 10 49 00 4a 00 bf af-01 00 48 00 00 00 ff 00 : ..I.J.....H.....
IFR 0x7faf0710 string 0x004b is "username"
IFR 0x7faf0710 string 0x004c is "User name"
IFR 0x7faf0710 string 0x004d is "http://ipxe.org/cfg/username"
IFR 0x7faf0710 string prompt 0x004c help 0x004d question 0xafbe varstore 0x0001:0x004b
000001db : 1c 10 4c 00 4d 00 be af-01 00 4b 00 00 00 ff 00 : ..L.M.....K.....
IFR 0x7faf0710 string 0x004e is "reverse-password"
IFR 0x7faf0710 string 0x004f is "Reverse password"
IFR 0x7faf0710 string 0x0050 is "http://ipxe.org/cfg/reverse-password"
IFR 0x7faf0710 string prompt 0x004f help 0x0050 question 0xafc1 varstore 0x0001:0x004e
000001eb : 1c 10 4f 00 50 00 c1 af-01 00 4e 00 00 00 ff 00 : ..O.P.....N.....
IFR 0x7faf0710 string 0x0051 is "reverse-username"
IFR 0x7faf0710 string 0x0052 is "Reverse user name"
IFR 0x7faf0710 string 0x0053 is "http://ipxe.org/cfg/reverse-username"
IFR 0x7faf0710 string prompt 0x0052 help 0x0053 question 0xafc0 varstore 0x0001:0x0051
000001fb : 1c 10 52 00 53 00 c0 af-01 00 51 00 00 00 ff 00 : ..R.S.....Q.....
IFR 0x7faf0710 string 0x0054 is "cert"
IFR 0x7faf0710 string 0x0055 is "Certificate"
IFR 0x7faf0710 string 0x0056 is "http://ipxe.org/cfg/cert"
IFR 0x7faf0710 string prompt 0x0055 help 0x0056 question 0xaf5b varstore 0x0001:0x0054
0000020b : 1c 10 55 00 56 00 5b af-01 00 54 00 00 00 ff 00 : ..U.V.[...T.....
IFR 0x7faf0710 string 0x0057 is "crosscert"
IFR 0x7faf0710 string 0x0058 is "Cross-signed certificate source"
IFR 0x7faf0710 string 0x0059 is "http://ipxe.org/cfg/crosscert"
IFR 0x7faf0710 string prompt 0x0058 help 0x0059 question 0xaf5d varstore 0x0001:0x0057
0000021b : 1c 10 58 00 59 00 5d af-01 00 57 00 00 00 ff 00 : ..X.Y.]...W.....
IFR 0x7faf0710 string 0x005a is "privkey"
IFR 0x7faf0710 string 0x005b is "Private key"
IFR 0x7faf0710 string 0x005c is "http://ipxe.org/cfg/privkey"
IFR 0x7faf0710 string prompt 0x005b help 0x005c question 0xaf5c varstore 0x0001:0x005a
0000022b : 1c 10 5b 00 5c 00 5c af-01 00 5a 00 00 00 ff 00 : ..[.\.\...Z.....
IFR 0x7faf0710 string 0x005d is "trust"
IFR 0x7faf0710 string 0x005e is "Trusted root certificate fingerprints"
IFR 0x7faf0710 string 0x005f is "http://ipxe.org/cfg/trust"
IFR 0x7faf0710 string prompt 0x005e help 0x005f question 0xaf5a varstore 0x0001:0x005d
0000023b : 1c 10 5e 00 5f 00 5a af-01 00 5d 00 00 00 ff 00 : ..^._.Z...].....
IFR 0x7faf0710 string 0x0060 is "dhcp-server"
IFR 0x7faf0710 string 0x0061 is "DHCP server"
IFR 0x7faf0710 string 0x0062 is "http://ipxe.org/cfg/dhcp-server"
IFR 0x7faf0710 string prompt 0x0061 help 0x0062 question 0x0036 varstore 0x0001:0x0060
0000024b : 1c 10 61 00 62 00 36 00-01 00 60 00 00 00 ff 00 : ..a.b.6...`.....
IFR 0x7faf0710 string 0x0063 is "priority"
IFR 0x7faf0710 string 0x0064 is "Settings priority"
IFR 0x7faf0710 string 0x0065 is "http://ipxe.org/cfg/priority"
IFR 0x7faf0710 string prompt 0x0064 help 0x0065 question 0xaf01 varstore 0x0001:0x0063
0000025b : 1c 10 64 00 65 00 01 af-01 00 63 00 00 00 ff 00 : ..d.e.....c.....
IFR 0x7faf0710 string 0x0066 is "scriptlet"
IFR 0x7faf0710 string 0x0067 is "Boot scriptlet"
IFR 0x7faf0710 string 0x0068 is "http://ipxe.org/cfg/scriptlet"
IFR 0x7faf0710 string prompt 0x0067 help 0x0068 question 0xaf51 varstore 0x0001:0x0066
0000026b : 1c 10 67 00 68 00 51 af-01 00 66 00 00 00 ff 00 : ..g.h.Q...f.....
IFR 0x7faf0710 end
0000027b : 29 02                                           : ).              
IFR 0x7faf0710 end
0000027d : 29 02                                           : ).              
IFR 0x7faf0710 string 0x0069 is "English"
<0-Y>EFIDEV PciRoot(0x0)/Pci(0x3,0x0)/MAC(5254FFFFEE1B,0x1) added child<0-Y> PciRoot(0x0)/Pci(0x3,0x0)/MAC(5254FFFFEE1B,0x1)/VenHw(92C0DA88-5296-EC5A-386C-8D88D5799C87)
<0-Y>SNPDEV 0x7db1a9c8 installed for net0 as device PciRoot(0x0)/Pci(0x3,0x0)/MAC(5254FFFFEE1B,0x1)
PRIVKEY has no private key
SNPDEV 0x7db1a9c8 link is up
EFIPCI 0000:00:03.0 using driver "virtio-net"
<0-Y>EFIDRV PciRoot(0x0)/Pci(0x3,0x0) using driver "PCI"
SNPDEV 0x7db1a9c8 GET_STATUS INTS:00
SNPDEV 0x7db1a9c8 START
SNPDEV 0x7db1a9c8 INITIALIZE (0 extra RX, 0 extra TX)
EFIPCI 0000:00:03.0 found range [c0000000,c0100000) -> [c0000000,c0100000)
EFIPCI 0000:00:03.0 found range [4800000000,4800100000) -> [4800000000,4800100000)
EFIPCI 0000:00:03.0 480000b000+2 -> 480000b000
EFIPCI 0000:00:03.0 found range [c0000000,c0100000) -> [c0000000,c0100000)
EFIPCI 0000:00:03.0 found range [4800000000,4800100000) -> [4800000000,4800100000)
EFIPCI 0000:00:03.0 480000b004+2 -> 480000b004
SNPDEV 0x7db1a9c8 link is up
SNPDEV 0x7db1a9c8 RECEIVE_FILTERS 00000005&~00000002 reset 0 mcast
SNPDEV 0x7db1a9c8 RECEIVE_FILTERS 00000005&~00000002 reset 0 mcast
SNPDEV 0x7db1a9c8 GET_STATUS!!!! X64 Exception Type - 0D(#GP - General Protection)  CPU Apic ID - 00000000 !!!!
ExceptionData - 0000000000000000
RIP  - 000000007DAD2A1E, CS  - 0000000000000038, RFLAGS - 0000000000010202
RAX  - B49D6968BF50B458, RCX - 0000000000000000, RDX - 0000000000000000
RBX  - 000000007DB1A6C8, RSP - 000000007FAF06C8, RBP - 0000000000000000
RSI  - B49D6968BF50B458, RDI - B49D6968BF50B458
R8   - 000000007FAF0368, R9  - 000000000001C200, R10 - 000000007DB9A690
R11  - 000000007DB1E032, R12 - 000000007DB1A7C8, R13 - 000000007DB1AE10
R14  - 000000007FB1C540, R15 - 000000007D2CD2AC
DS   - 0000000000000030, ES  - 0000000000000030, FS  - 0000000000000030
GS   - 0000000000000030, SS  - 0000000000000030
CR0  - 0000000080010033, CR2 - 0000000000000000, CR3 - 000000007F601000
CR4  - 0000000000000668, CR8 - 0000000000000000
DR0  - 0000000000000000, DR1 - 0000000000000000, DR2 - 0000000000000000
DR3  - 0000000000000000, DR6 - 00000000FFFF0FF0, DR7 - 0000000000000400
GDTR - 000000007F3DE000 0000000000000047, LDTR - 0000000000000000
IDTR - 000000007EE8E018 0000000000000FFF,   TR - 0000000000000000
FXSAVE_STATE - 000000007FAF0320
!!!! Find image based on IP(0x7DAD2A1E) 1af41000.efidrv (ImageBase=000000007DAB8000, EntryPoint=000000007DAC9A08) !!!!


If SEV is disabled, things work (i.e. it boots up fine).
The lines in the iPXE DEBUG output at the same spot WITHOUT SEV is:
SNPDEV 0x7df549c8 link is up
SNPDEV 0x7df549c8 RECEIVE_FILTERS 00000005&~00000002 reset 0 mcast
SNPDEV 0x7df549c8 RECEIVE_FILTERS 00000005&~00000002 reset 0 mcast
SNPDEV 0x7df549c8 GET_STATUS INTS:01    <----- crash occurs here with SEV (before we get to the "INTS:01" output)
SNPDEV 0x7df549c8 RECEIVE 0x7d93329e(+5f2)+3e
SNPDEV 0x7df549c8 RECEIVE_FILTERS 0000000d&~0000000a 0 mcast
SNPDEV 0x7df549c8 GET_STATUS INTS:01
SNPDEV 0x7df549c8 RECEIVE 0x7d93201e(+5f2)+3e
SNPDEV 0x7df549c8 RECEIVE 0x7d93269e(+5f2)+3e
SNPDEV 0x7df549c8 GET_STATUS INTS:01

**Comment by mcb30** [comment]
Does it work if you use a different virtual NIC instead of virtio?  It's possible that SEV requires drivers to use the DMA mapping API (rather than assuming a flat memory map accessible by all DMA-capable hardware).  The iPXE Intel and Realtek drivers have been updated to use the DMA mapping API, but not the virtio driver.

**Reply by ajyoung-oracle** [reply]
No, we haven't tested other than virtio. It looks to me like it's crashing somewhere in efi_snp_poll() the very first time there is a receive packet to process. The key thing to note about SEV is that a VM runs with encypted memory (which is tranparent to a guest)  *except* for DMA buffers - which need to be marked unencrypted (when SEV is enabled). Thus entities which use DMA buffers need to be SEV aware I believe, There was a special protocol added to EDK2 (gEdkiiIoMmuProtocolGuid) for this purpose. You can see the changes made to use this new protocol (when SEV is enabled) in OvmfPkg/Library/QemuFwCfgLib/. Compare the implementation of InternalQemuFwCfgDmaBytes() at commit level fee47a261c65235e263aa128d28e7da8f8f9f100 (prior to SEV support) to the most recent. This clearly show how the new protocol is used. My suspicion is we need something similar done for iPXE virtio.

**Reply by ajyoung-oracle** [reply]
Changes to add the new protocol were also made to PciHostBridgeDxe (See PciIo.c:PciIoMap() for example). So, I would think any code calling PciIo->Map() should work too. I need to ask about this.

**Reply by mcb30** [reply]
The iPXE DMA mapping API (i.e. `dma_alloc()`, `dma_map()`, `dma()`, `iob_map_tx()`, `iob_map_rx()`, etc) provides an abstraction that will, on a UEFI system with a PCI device, end up calling `PciIo->Map()`.  So, it is very likely that updating the iPXE virtio-net driver to use this API will fix the problem.

**Reply by mcb30** [reply]
The commit log for https://github.com/ipxe/ipxe/commit/8d337ecda includes a good description of the steps required to update a driver to use the DMA API.

**Reply by ajyoung-oracle** [reply]
OK. Thanks. I will take a look into that.

**Comment by ajyoung-oracle** [comment]
Update: I updated the iPXE virtio-net driver to use the new DMA API and indeed it now functions with AMD SEV enabled (instead of crashing as above).

I submitted a PR containing the new code: https://github.com/ipxe/ipxe/pull/505

**Reply by mcb30** [reply]
PR #505 merged - this should now work as expected.  Thank you!
